### DISCUSSION TITLE
appSettings e Configuration

### DISCUSSION THREAD

**Discussion by rafaferreirasantos** [balta-io/2811#17] [Q&A] [answered]
Senhores,

Verifiquei que o carregamento das informações do appSettings para a classe Configuration não está funcionando corretamente.
**Se pegar o código do repositório do curso e apagar o valor das variáveis da classe Configuration o código para de funcionar**, pois ao rodar a parte de configuração de autenticação ele ainda não carregou os dados do appSettings na classe Configuration. 
Tentei então alterar a ordem em que é chamado o ConfigureAuthentication, mas aí eu recebo o erro: "Cannot modify ServiceCollection after application is built."

Podem auxiliar em como remover as credenciais do código e ler do appSettings?

Obrigado.

**Comment by andrebaltieri** [comment]
Bom dia @zecaloteiro, você chegou até o último módulo do curso onde mostro a diferença entre produção e dev, configuramos as connection strings e tudo mais? Seria o módulo 8!

**Reply by rafaferreirasantos** [reply]
Fala Balta. Cheguei sim até a última aula do módulo 8. Mas em nenhuma aula você apaga as "credenciais" do arquivo Configuration.cs tanto que no repositório está o arquivo está "commitado" com as informações... e só por esse motivo que o código funciona. 

Se você apagar o valor das variáveis JwtKey, ApiKeyName e ApiKey de dentro do arquivo Configuration.cs e tentar rodar seu código (seja em modo release ou debug) a aplicação não vai funcionar.

Pra fazer o teste eu criei as variáveis no appsettings.Development.json e rodei em modo debug, e não funciona.

Se puder por favor fazer o teste, agradeço.

Só lembrando novamente que, fiz o teste com  o código que está no repositório pra não correr o risco de ter errado alguma coisa por aqui.

Teste simples: 
1) Baixar o código do repositório do Módulo 8;
2) Apagar o valor das variáveis do arquivo Configuration.cs
3) Criar as variáveis no appsettings.Development.json 
4) Rodar em modo debug/development.
A aplicação não roda.
![Screenshot_1](https://user-images.githubusercontent.com/7022211/160424243-7d588914-fa30-4ba6-8326-6aaf57f1e3f9.png)

**Reply by andrebaltieri** [reply]
Vou revisar, mas possívelmente isto ocorre por que as configurações do JWT estão send carregadas depois do `AddJwtBearer` logo a chave não existe neste momento.

Precisa carregar as informações do `appSettings.json` antes de aplicar as configurações.

**Reply by marianohexsel** [reply]
Oi Balta, estou no mesmo ponto que o colega zecaloteiro.
Identifiquei a mesma coisa, agora precisamos saber como fazemos para carregar antes, aparentemente precisa do Builder na Autenticação (onde usa o Key do Jwt) e no Services, para o TokenService.

Pra gerar o App precisa do Builder, mas é com o App que busca as informações do arquivo "appsettings.json"

**Comment by gabrielhs1357** [comment, marked as answer]
Olá galera, estou tendo o mesmo problema por aqui.

Só complementando, o método ConfigureAuthentication(builder) usa a chave JWTKey mas esse valor ainda não foi lido do app settings porque essa leitura só é feita no método LoadConfiguration(app).

Aqui está o erro quando tento rodar o projeto:
 
![1](https://user-images.githubusercontent.com/50048495/164943574-beb5204c-1b3e-4e13-953f-b745f427f801.png)

Ao tentar inverter os métodos e chamar o LoadConfiguration antes do ConfigureAuthentication, obtenho um erro dizendo que não é possível modificar as services depois que o app foi buildado: "Cannot modify ServiceCollection after application is built.".

**Reply by andrebaltieri** [reply]
Oi pessoal, boa noite... eu vou mostrar isto nas próximas aulas, mas aqui vai a solução...

Conforme eu comento nas aulas de configuração, tanto `app.Configuration` quanto `builder.Configuration` existem... então se vocês precisam obter uma configuração em tempo de build, tem que usar o `builder.Configuration`.

```csharp
var builder = WebApplication.CreateBuilder(args);
LoadConfiguration(builder); // Mudar para este ponto
ConfigureAuthentication(builder);
ConfigureMvc(builder);
ConfigureServices(builder);

builder.Services.AddEndpointsApiExplorer();
builder.Services.AddSwaggerGen();

var app = builder.Build();
...

// Agora ao invés de usar app, usamos o builder
void LoadConfiguration(WebApplicationBuilder builder)
{
}
```

O código final fica assim:

```csharp
using System.IO.Compression;
using System.Text;
using System.Text.Json.Serialization;
using Blog;
using Blog.Data;
using Blog.Services;
using Microsoft.AspNetCore.Authentication.JwtBearer;
using Microsoft.AspNetCore.ResponseCompression;
using Microsoft.EntityFrameworkCore;
using Microsoft.IdentityModel.Tokens;

var builder = WebApplication.CreateBuilder(args);
LoadConfiguration(builder);
ConfigureAuthentication(builder);
ConfigureMvc(builder);
ConfigureServices(builder);

builder.Services.AddEndpointsApiExplorer();
builder.Services.AddSwaggerGen();

var app = builder.Build();
app.UseHttpsRedirection();
app.UseAuthentication();
app.UseAuthorization();
app.MapControllers();
app.UseStaticFiles();
app.UseResponseCompression();

if (app.Environment.IsDevelopment())
{
    app.UseSwagger();
    app.UseSwaggerUI();
}

app.Run();

void LoadConfiguration(WebApplicationBuilder builder)
{
    Configuration.JwtKey = builder.Configuration.GetValue<string>("JwtKey");
    Configuration.ApiKeyName = builder.Configuration.GetValue<string>("ApiKeyName");
    Configuration.ApiKey = builder.Configuration.GetValue<string>("ApiKey");

    var smtp = new Configuration.SmtpConfiguration();
    builder.Configuration.GetSection("SmtpConfiguration").Bind(smtp);
    Configuration.Smtp = smtp;
}

void ConfigureAuthentication(WebApplicationBuilder builder)
{
    var key = Encoding.ASCII.GetBytes(Configuration.JwtKey);
    builder.Services.AddAuthentication(x =>
    {
        x.DefaultAuthenticateScheme = JwtBearerDefaults.AuthenticationScheme;
        x.DefaultChallengeScheme = JwtBearerDefaults.AuthenticationScheme;
    }).AddJwtBearer(x =>
    {
        x.TokenValidationParameters = new TokenValidationParameters
        {
            ValidateIssuerSigningKey = true,
            IssuerSigningKey = new SymmetricSecurityKey(key),
            ValidateIssuer = false,
            ValidateAudience = false
        };
    });
}

void ConfigureMvc(WebApplicationBuilder builder)
{
    builder.Services.AddMemoryCache();
    builder.Services.AddResponseCompression(options =>
    {
        // options.Providers.Add<BrotliCompressionProvider>();
        options.Providers.Add<GzipCompressionProvider>();
        // options.Providers.Add<CustomCompressionProvider>();
    });
    builder.Services.Configure<GzipCompressionProviderOptions>(options =>
    {
        options.Level = CompressionLevel.Optimal;
    });
    builder
        .Services
        .AddControllers()
        .ConfigureApiBehaviorOptions(options => { options.SuppressModelStateInvalidFilter = true; })
        .AddJsonOptions(x =>
        {
            x.JsonSerializerOptions.ReferenceHandler = ReferenceHandler.IgnoreCycles;
            x.JsonSerializerOptions.DefaultIgnoreCondition = JsonIgnoreCondition.WhenWritingDefault;
        });
}

void ConfigureServices(WebApplicationBuilder builder)
{
    var connectionString = builder.Configuration.GetConnectionString("DefaultConnection");
    builder.Services.AddDbContext<BlogDataContext>(
        options =>
            options.UseSqlServer(connectionString));
    builder.Services.AddTransient<TokenService>();
    builder.Services.AddTransient<EmailService>();
}
```
